hpx/parallel/algorithms/reduce_by_key.hpp

```diff
@@ -0,0 +1,561 @@
+    ///                     will not apply any non-constant function through the
+    ///                     dereferenced iterator.
+    ///
+    /// \a comp has to induce a strict weak ordering on the values.
+    ///
+    /// The application of function objects in parallel algorithm
+    /// invoked with an execution policy object of type
+    /// \a sequential_execution_policy execute in sequential order in the
+    /// calling thread.
+    ///
+    /// The application of function objects in parallel algorithm
+    /// invoked with an execution policy object of type
+    /// \a parallel_execution_policy or \a parallel_task_execution_policy are
+    /// permitted to execute in an unordered fashion in unspecified
+    /// threads, and indeterminately sequenced within each thread.
+    ///
+    /// \returns  The \a reduce_by_key algorithm returns a
+    ///           \a hpx::future<pair<Iter1,Iter2>> if the execution policy is of
+    ///           type
+    ///           \a sequential_task_execution_policy or
+    ///           \a parallel_task_execution_policy and returns \a pair<Iter1,Iter2>
+    ///           otherwise.
+    //-----------------------------------------------------------------------------
+
+    template<
+        typename ExPolicy,
+        typename RanIter, typename RanIter2, typename OutIter, typename OutIter2,
+        typename Compare =
+            std::equal_to<typename std::iterator_traits<RanIter>::value_type>,
+        HPX_CONCEPT_REQUIRES_(is_execution_policy<ExPolicy>::value &&
+            hpx::traits::is_iterator<RanIter>::value &&
+            hpx::traits::is_iterator<RanIter2>::value &&
+            hpx::traits::is_iterator<OutIter>::value &&
+            hpx::traits::is_iterator<OutIter2>::value
+        )
+    >
+
+    typename util::detail::algorithm_result<ExPolicy, std::pair<OutIter, OutIter2>>::type
+        reduce_by_key(ExPolicy &&policy, RanIter key_first, RanIter key_last,
+            RanIter2 values_first, OutIter keys_output, OutIter2 values_output,
+            Compare &&comp = Compare())
+    {
+        typedef util::detail::algorithm_result<
+            ExPolicy, std::pair<OutIter, OutIter2>> result;
+        typedef typename std::iterator_traits<RanIter>::iterator_category  category1;
+        typedef typename std::iterator_traits<RanIter2>::iterator_category category2;
+        typedef typename std::iterator_traits<OutIter>::iterator_category  category3;
+        typedef typename std::iterator_traits<OutIter2>::iterator_category category4;
+
+        static_assert(
+            (boost::is_base_of<std::random_access_iterator_tag, category1>::value) ||
+            (boost::is_base_of<std::random_access_iterator_tag, category2>::value) ||
+            (boost::is_base_of<std::output_iterator_tag, category3>::value) ||
+            (boost::is_base_of<std::output_iterator_tag, category4>::value),
+            "iterators : Random_access for inputs and Output for outputs.");
```

Comment:
Did you intend to use `&&` instead of `||`? Also, any iterator type except input_iterators can be used as an output iterator. What you probably want to write in order to check whether an iterator conforms to the output iterator concept is:

    hpx::traits::is_output_iterator<OutIter>::value || 
        hpx::traits::is_forward_iterator<OutIter>::value 

We also have a 

    hpx::traits::is_random_access_iterator<RanIter>::value

which could be used instead (this also gets rid of the boost/type_traits facilities which we replaced everywhere else in the parallelism files)..